### PR TITLE
Add parameters.resolveIncludes and parameters.resolveExcludes for RESOLVE_ONLY zip repo generation strategy

### PR DESCRIPTION
**Pull request by ppalaga**
### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/bacon/wiki/Changelog) for your change if user-facing?

It does not look like I have enough permission to edit the wiki

* [x] Have you added unit tests for your change?

### CODE REVIEW

**Comment by ppalaga** [comment]
@michalszynkiewicz @cunningt @harshmadhani @svkcemk @aloubyansky could you please review?

**Comment by ppalaga** [comment]
For the record, I was testing with this config for CEQ:

```
product:
  name: Camel Extensions for Quarkus
  abbreviation: camel-quarkus
  stage: AR3
  issueTrackerUrl: http://issues.jboss.org/browse/ENTESB
version: 2.2.1
milestone: AR3
group: Camel Extensions for Quarkus 2.2.1 AR3
defaultBuildParameters:
  environmentId: 21
  buildScript: 'mvn -e -V -B -DskipTests -DfailIfNoTests=false -Dtest=false clean
    deploy'

builds:
- name: camel-3.11.5-2.2.1
  project: jboss-fuse/camel
    #fetchScmUrl: https%3A%2F%2Fgithub.com%2Fjboss-fuse%2Fcamel.git
    #pushScmUrl: https%3A%2F%2Fgithub.com%2Fjboss-fuse%2Fcamel.git
    #tag: camel-3.11.5.fuse-800012
    #buildCommand: mvn+-e+-V+-B+-DskipTests+-DfailIfNoTests%3Dfalse+-Dtest%3Dfalse+-Penable-schemagen%2Crelease%2Capt+clean+install+-Dmaven.javadoc.skip%3Dtrue+-Dcamel.javadoc.offline%3Dtrue+-Dmaven.javadoc.isOffline%3Dtrue+-DkeepStagingRepositoryOnCloseRuleFailure%3Dtrue+
  scmUrl: git+ssh://code.engineering.redhat.com/jboss-fuse/camel.git
  externalScmUrl: ssh://git@github.com/jboss-fuse/camel.git
  scmRevision: camel-3.11.5.fuse-800012
  environmentId: 110
  pigYamlMetadata: Fuse dev build https%3A%2F%2Fci-jenkins-csb-fuse.apps.ocp-c1.prod.psi.redhat.com%2Fjob%2Fintegration.next%2Fjob%2Fcamel-quarkus-2.2.1%2F49%2F Fuse camelquarkus yaml config camel-quarkus-2.2.1.0.camelquarkus-800022
  buildPodMemory: 12
  buildScript: >
    export JAVA_TOOL_OPTIONS="-Xmx3g";
    export MAVEN_OPTS="-Xmx4g";
    mvn -e -V -B -DskipTests -Dmaven.compiler.testTarget=1.8 -Dcq.onCheckFailure=WARN -Dcq.camel-prod-excludes.skip=true -Dmaven.compiler.testSource=1.8 -DtarLongFileMode=posix -DfailIfNoTests=false -Dtest=false -Dcamel.javadoc.offline=true -Dmaven.javadoc.isOffline=true -Penable-schemagen,release,apt
    clean deploy -DkeepStagingRepositoryOnCloseRuleFailure=true -Dgpg.skip=true dependency:tree
  alignmentParameters:
    - -DprojectMetaSkip=true -DdependencyOverride.com.google.protobuf:*@*= -DdependencyOverride.org.mongodb:*@*= -DdependencyOverride.org.apache.tomcat:*@*= -DdependencyOverride.com.sun.xml.bind:*@*= -DdependencyOverride.com.github.everit-org.json-schema:*@*=1.6.0.redhat-1 -DdependencyOverride.org.codehaus.woodstox:*@*= -DdependencyOverride.org.apache.abdera:*@*= -DdependencyOverride.org.reactivestreams:*@*= -DdependencyOverride.org.bouncycastle:*@*= -DdependencyOverride.org.apache.kafka:*@*= -DdependencyOverride.javax.ws.rs:*@*=2.1.5 -DdependencyOverride.io.reactivex.rxjava2:*@*= -DdependencyOverride.io.reactivex:*@*= -DdependencyOverride.org.apache.activemq:*@*= -DdependencyOverride.com.github.ben-manes.caffeine:caffeine:*@*= -DdependencyOverride.commons-codec:commons-codec:*@*= -DdependencyOverride.org.codehaus.jettison:jettison:*@*= -DdependencyOverride.org.jdom:jdom:*@*= -DdependencyOverride.com.netflix.hystrix:*@*= -DdependencyOverride.javax.xml.bind:*@*= -DdependencyOverride.org.glassfish:*@*=  -DdependencyOverride.javax.el:*@*= -DdependencyOverride.org.glassfish.jersey.core:*@*= -DdependencyOverride.org.glassfish.jersey.media:*@*= -DdependencyOverride.org.glassfish.jersey.inject:*@*= -DdependencyOverride.org.glassfish.jersey.containers:*@*= -DdependencyOverride.org.infinispan:*@*= -DdependencyOverride.org.infinispan.server:*@*= -DdependencyOverride.io.grpc:*@*= -DdependencyOverride.jakarta.xml.bind:*@*=
  extraRepositories:
    - http://resources.knopflerfish.org/repo/maven2/release/
    - http://packages.confluent.io/maven/
    - http://maven.icm.edu.pl/artifactory/repo/

- name: camel-quarkus-2.2.1-2.2.1
    #fetchScmUrl: ${camel-quarkus.fetch.scmUrl}
    #pushScmUrl: ${camel-quarkus.push.scmUrl}
    #tag: camel-quarkus-2.2.1.fuse-800015
    #buildCommand: ${camel-quarkus.buildCommand}
  project: jboss-fuse/camel-quarkus
  scmUrl: git+ssh://code.engineering.redhat.com/jboss-fuse/camel-quarkus.git
  externalScmUrl: ssh://git@github.com/jboss-fuse/camel-quarkus.git
  scmRevision: camel-quarkus-2.2.1.fuse-800015
  environmentId: 240
  pigYamlMetadata: Fuse dev build https%3A%2F%2Fci-jenkins-csb-fuse.apps.ocp-c1.prod.psi.redhat.com%2Fjob%2Fintegration.next%2Fjob%2Fcamel-quarkus-2.2.1%2F49%2F Fuse camelquarkus yaml config camel-quarkus-2.2.1.0.camelquarkus-800022
  buildPodMemory: 12
  buildScript: >
    export MAVEN_OPTS="-Xmx8000m"

    mvn -B -Pdeploy clean deploy -Dcq.prod-artifacts.skip=true -Denforcer.skip=true -DskipTests -DaltDeploymentRepository=indy-mvn::${AProxDeployUrl} dependency:tree
  alignmentParameters:
    - -DscanActiveProfiles=true -DstrictPropertyValidation=false -DgroovyScripts=https://gitlab.cee.redhat.com/tcunning/cq-pre-pme-manipulator/-/raw/main/src/main/groovy/com/redhat/fuse/groovy/pmeCamelQuarkusPrePMEManipulator.groovy,https://gitlab.cee.redhat.com/tcunning/camel-quarkus-pom-manipulator/raw/master/src/main/groovy/com/redhat/fuse/groovy/pmeAlterCamelQuarkus.groovy
  dependencies:
    - camel-3.11.5-2.2.1

outputPrefixes:
  releaseFile: rhi-camel-extensions-for-quarkus
  releaseDir: rhi-camel-extensions-for-quarkus

flow:
  licensesGeneration:
    strategy: GENERATE

  repositoryGeneration:
    strategy: RESOLVE_ONLY
    sourceBuild: camel-quarkus-2.2.1-2.2.1
    sourceArtifact: 'camel-quarkus-bom-[\d]+.*.pom'
    parameters:
       suspectArtifacts: "javax.annotation:javax.annotation-api,javax.enterprise:cdi-api"
       #extensionsListUrl: "https://gitlab.cee.redhat.com/soghosh/build-configurations/-/raw/camel-quarkus/camel-quarkus/2.2.1/extensionArtifactList.txt"
       resolveIncludes: "*:*:*redhat-*"
       resolveExcludes: "io.netty:*:*:linux-aarch_64:*,io.netty:*:*:osx-*:*,io.smallrye.reactive:vertx-mutiny-clients,io.vertx:vertx-sql-client-parent,io.vertx:vertx-grpc-parent,io.vertx:vertx-web:js:client:*,io.quarkus:quarkus-oidc-client-filter*,io.quarkus:quarkus-rest-client*"
    externalAdditionalArtifacts:
      - 'io.quarkus:quarkus-maven-plugin:pom:2.2.5.Final-redhat-00010'
      - 'io.quarkus:quarkus-maven-plugin:pom:2.2.5.Final-redhat-00010'
#addons:
#  notYetAlignedFromDependencyTree:
#  offlineManifestGenerator:
#    offlineManifestFileName: rhi-camel-extensions-for-quarkus-offliner-2.2.1.txt
```

**Comment by svkcemk** [comment]
Looking good to me!

**Comment by ppalaga** [comment]
It would be easier not forget to document the new options if the docs lived in the same branch as the code.

**Comment by ppalaga** [comment]
@michalszynkiewicz any idea whether the test failures have anything to do with the current changes?

**Comment by ppalaga** [comment]
Here is the docs: https://github.com/project-ncl/bacon/pull/920

**Comment by ppalaga** [comment]
cc @paulgallagher75

**Comment by michalszynkiewicz** [comment]
> @michalszynkiewicz any idea whether the test failures have anything to do with the current changes?

no idea, it's been a while since I did anything in bacon.

**Comment by ppalaga** [comment]
> no idea, it's been a while since I did anything in bacon.

Who else could tell then?

**Comment by rnc** [comment]
> The CI job returns 404 for me now. @thescouser89 is there any way to restart the job?

@jomrazek  @thescouser89  Can we just move to GitHub Actions?!

**Comment by thescouser89** [comment]
retest this please

**Comment by thescouser89** [comment]
@ppalaga  It's failing due to formatting issues. We try to keep our codebase formatting consistent. Please run:
```
mvn clean install -DskipTests=true
```
and commit or amend your commit to include the formatting changes needed. Thanks!

**Comment by ppalaga** [comment]
> It's failing due to formatting issues.

Oh, thanks for the hint. Done in d741d12

**Comment by thescouser89** [comment]
retest this please

**Comment by thescouser89** [comment]
retest this please

**Comment by thescouser89** [comment]
retest this please

**Comment by thescouser89** [comment]
So sorry for the delay. It's a combination of laziness and some CI issues.

**Comment by thescouser89** [comment]
For some reason, those changes break slf4j classloading. I get:
```
/projects/work/bacon ppalaga-220322-resolveIncludes !1 ❯ java -jar ~/projects/work/bacon/cli/target/bacon.jar -v pnc -o build get 0                                                     1m 2s
Could not invoke public static void org.jboss.pnc.bacon.cli.App.setVerbosityIfPresent(boolean) with true (java.lang.ClassCastException: class org.slf4j.impl.Slf4jLogger cannot be cast to class ch.qos.logback.classic.Logger (org.slf4j.impl.Slf4jLogger and ch.qos.logback.classic.Logger are in unnamed module of loader 'app'))

```

Checking why this happened.

**Comment by ppalaga** [comment]
> For some reason, those changes break slf4j classloading

OK, I took a more defensive strategy to avoid getting incompatible Maven artifacts in 8bd5d58

Besides, I took care to make all tests work locally. I hope they'll pass on the CI too. 

Thanks for your patience!

**Comment by thescouser89** [comment]
Yup on it!

**Comment by thescouser89** [comment]
LGTM! Thanks for adding the tests for `GAVSet` and `GAVPattern` which helped me understand how it is used in `RepoManager`!

**Comment by ppalaga** [comment]
Thanks a lot @thescouser89! When can we expect a release?

**Comment by thescouser89** [comment]
Oh we are getting a release on Tuesday that won't include that fix. but we can cut another one just afterwards if needed.

CC: @jomrazek would you happen to know?

**Comment by jomrazek** [comment]
we can release this with Jan's improvement sometime next week when Jan's change is tested properly

**Comment by ppalaga** [comment]
> we can release this with Jan's improvement sometime next week when Jan's change is tested properly

That would be great, thanks!

**Comment by ppalaga** [comment]
Any update about the release, @jomrazek ?

**Comment by thescouser89** [comment]
@ppalaga Jozef released the 2.2.1 Bacon version with your PR and it should now be available. To upload to the latest one, either download it via here: https://repo1.maven.org/maven2/org/jboss/pnc/bacon/cli/2.2.1/cli-2.2.1-shaded.jar or by running:
```
$ bacon update
```

**Comment by ppalaga** [comment]
Thanks a lot for the news and for the release!

File: pom.xml [hunk 1]
```diff
@@ -402,6 +403,66 @@
...
+                        <artifactId>javax.inject</artifactId>
+                    </exclusion>
+                    <exclusion>
+                        <groupId>org.slf4j</groupId>
+                        <artifactId>slf4j-api</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
+            <dependency>
+                <groupId>org.apache.maven</groupId>
```
Comment (jomrazek) [inline pom.xml:427]: @ppalaga hi is this dependency used somewhere and needed? It broke license generation, I though about updating licenses-generator but looking at it, embedder is used in it as some jenkins library which seems obsolete and I'm not feeling like digging in that. In short i removed it, tests are passing and my test builds with generation tests are working too.

Comment (jomrazek) [reply pom.xml:427]: (Also I downgraded mvn to 3.6.3 because there was change in 3.8.x)

**Pull request merged**